### DISCUSSION TITLE
basic image shell line-wrapping / command-recall unusable

### DISCUSSION THREAD

**Discussion by cevich** [devcontainers/spec#456] [Q&A]
I'm not able to use Docker, so my container engine is podman (`v4.9.3`).  I'm using a very basic devcontainer through the CLI.  The `up` and `exec` commands seem to work fine, and I'm able to run specific commands in my devcontainer without any problems.

```json
{
    "name": "basic",
    "image": "mcr.microsoft.com/vscode/devcontainers/base:debian-12",
    "features": {
    },
    "securityOpt": [
        "label=disable",
        "apparmor=unconfined"
    ],
    "runArgs": [
        "--userns=keep-id"
    ]
}
```

**The issue is**, if I `exec` the `bash` (or `zsh`) command, the terminal prompt's line-wrapping and command-recall functions misbehave and render the console unusable.

I'm able to reproduce the problem with the `mcr.microsoft.com/vscode/devcontainers/base:ubuntu-22.04` image as well.  I've tried within the GNOME Terminal app as well as directly on a Linux console (i.e. ctrl-alt-F3).  I'm even able to reproduce directly using the image:

```
podman run -it --rm mcr.microsoft.com/vscode/devcontainers/base:debian-12 bash
root ➜ / $ cd /usr/local/bin
he next linelocal/bin $ # this is a really long command that should wrap to t 
```

As you can see, it's wrapping ONTO the current terminal line, not a new one.  If I press the up-arrow (command-recall) I get a similar effect.  After a series of badly wrapped lines, this "off-by-one line" problem seems to stack, jumping the cursor up an extra line each time, rendering the terminal unusable.

Has anyone else experienced this behavior and/or know of a way to fix it?

Note: Inside the devcontainer or when directly running the base image, the system is aware of the terminal size, including when I change it:

```
root ➜ /usr/local/bin $ tput cols
79
root ➜ /usr/local/bin $ tput cols
80
root ➜ /usr/local/bin $ echo $COLUMNS
80
e next line/local/bin $ # this is a really long command that should wrap to the
```

**Comment by samruddhikhandale** [comment]
Hi 👋 

This issue seems to be related to the terminal settings and not specifically to the programming or the dev container setup. However, you can try to force the terminal to use the correct settings by adding the following line to your `.bashrc` or `.zshrc` file:

```
export TERM=xterm
```

This line sets the TERM environment variable to xterm, which is a terminal type that supports line wrapping and command recall. After adding this line, you need to source the .bashrc or .zshrc file to apply the changes. If this doesn't solve the issue, it might be a bug in the terminal emulator or in the Podman container engine. In that case, you should report the issue to the maintainers of the software.

**Reply by cevich** [reply]
When I run bash in `mcr.microsoft.com/vscode/devcontainers/base:debian-12` image it reports `$TERM` is already set to `xterm`.  I think the problem is a bug w/in the image somewhere, any idea where it was built from (i.e. `Dockerfile`)?

**Comment by mohshami** [comment]
I've been trying to figure this out all day today, and I think I know the cause

in /home/vscode/.bashrc
```
# bash theme - partly inspired by https://github.com/ohmyzsh/ohmyzsh/blob/master/themes/robbyrussell.zsh-theme
__bash_prompt() {
    local userpart='`export XIT=$? \
        && [ ! -z "${GITHUB_USER:-}" ] && echo -n "\[\033[0;32m\]@${GITHUB_USER:-} " || echo -n "\[\033[0;32m\]\u " \
        && [ "$XIT" -ne "0" ] && echo -n "\[\033[1;31m\]➜" || echo -n "\[\033[0m\]➜"`'
    local gitbranch='`\
        if [ "$(git config --get devcontainers-theme.hide-status 2>/dev/null)" != 1 ] && [ "$(git config --get codespaces-theme.hide-status 2>/dev/null)" != 1 ]; then \
            export BRANCH="$(git --no-optional-locks symbolic-ref --short HEAD 2>/dev/null || git --no-optional-locks rev-parse --short HEAD 2>/dev/null)"; \
            if [ "${BRANCH:-}" != "" ]; then \
                echo -n "\[\033[0;36m\](\[\033[1;31m\]${BRANCH:-}" \
                && if [ "$(git config --get devcontainers-theme.show-dirty 2>/dev/null)" = 1 ] && \
                    git --no-optional-locks ls-files --error-unmatch -m --directory --no-empty-directory -o --exclude-standard ":/*" > /dev/null 2>&1; then \
                        echo -n " \[\033[1;33m\]✗"; \
                fi \
                && echo -n "\[\033[0;36m\]) "; \
            fi; \
        fi`'
    local lightblue='\[\033[1;34m\]'
    local removecolor='\[\033[0m\]'
    PS1="${userpart} ${lightblue}\w ${gitbranch}${removecolor}\$ "
    unset -f __bash_prompt
}
__bash_prompt
export PROMPT_DIRTRIM=4
```

Running the following command disables the theme and therefore fixes the issue.

```
sed -i '/^__bash_prompt$/d' /home/vscode/.bashrc
```

You can also fix the issue by switching to a different shell

**Comment by haticecan37** [comment]
{
    "name": "basic",
    "image": "mcr.microsoft.com/vscode/devcontainers/base:debian-12",
    "features": {
    },
    "securityOpt": [
        "label=disable",
        "apparmor=unconfined"
    ],
    "runArgs": [
        "--userns=keep-id"
    ]
}
